### DISCUSSION TITLE
Reference for Rust mechanism

### DISCUSSION THREAD

**Discussion by alissa-tung** [rusterlium/rustler#464] [Q&A] [answered]
Hello, I am trying to use the rustler library to write a niflib for Erlang. I am wondering if there exists some documentation or reference about the following topics:

- the `lib` attributes used in Cargo manifest
- parametric polymorphic functions
- how lifetime and ownership, borrow checker work in nif

thx a lot

**Comment by evnu** [comment, marked as answer]
We don't have documentation on using Rustler from Erlang (yet). With regard to your questions:

> * the lib attributes used in Cargo manifest

I guess you mean the `crate-type` configuration? Our template (`rustler_mix/priv/templates/basic/Cargo.toml.eex `) defines `cdylib`. If I remember correctly, we need to do this as cdylibs are used when compiling into a shared library to be used from C/C++.

> * parametric polymorphic functions

Can you further explain what you mean? For a NIF, polymorphism is implemented by dynamic dispatch on the terms passed into the NIF. NIFs itself cannot be monomorphized, as C does not have that functionality.

> * how lifetime and ownership, borrow checker work in nif

This might be a large topic. Do you have a more specific question?

**Reply by alissa-tung** [reply]
Thank you for answering this!

I feel very sorry for the late of reply on this. At that time I was too busy at work so that I have not notified that you have answered the questions. At that time I have developed a library using Rustler library to expose APIs to Erlang. If someone also runs into this questions, I hope the links might be helpful:

- https://github.com/hstreamdb/hstreamdb_erl-rs
- https://github.com/hstreamdb/hstreamdb-rust/tree/main/src/x/hstreamdb-erl-nifs
